junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertIterableEquals.java

```diff
@@ -69,24 +79,39 @@ private static void assertIterableEquals(Iterable<?> expected, Iterable<?> actua
 			}
 
 			indexes.addLast(processed - 1);
-			assertIterableElementsEqual(expectedElement, actualElement, indexes, messageOrSupplier);
+			assertIterableElementsEqual(expectedElement, actualElement, indexes, messageOrSupplier,
+				visitedIterablesExpected, visitedIterablesActual);
 			indexes.removeLast();
 		}
 
 		assertIteratorsAreEmpty(expectedIterator, actualIterator, processed, indexes, messageOrSupplier);
 	}
 
 	private static void assertIterableElementsEqual(Object expected, Object actual, Deque<Integer> indexes,
-			Object messageOrSupplier) {
+			Object messageOrSupplier, Collection<Iterable<?>> visitedIterablesExpected,
+			Collection<Iterable<?>> visitedIterablesActual) {
 		if (expected instanceof Iterable && actual instanceof Iterable) {
-			assertIterableEquals((Iterable<?>) expected, (Iterable<?>) actual, indexes, messageOrSupplier);
+			assertIterableEquals((Iterable<?>) expected, (Iterable<?>) actual, indexes, messageOrSupplier,
+				visitedIterablesExpected, visitedIterablesActual);
 		}
 		else if (!Objects.equals(expected, actual)) {
 			assertIterablesNotNull(expected, actual, indexes, messageOrSupplier);
 			failIterablesNotEqual(expected, actual, indexes, messageOrSupplier);
 		}
 	}
 
+	private static void assertIterablesNotAlreadyVisited(Iterable<?> expected, Iterable<?> actual,
+			Deque<Integer> indexes, Object messageOrSupplier, Collection<Iterable<?>> visitedIterablesExpected,
+			Collection<Iterable<?>> visitedIterablesActual) {
+		if (visitedIterablesExpected.contains(expected) || visitedIterablesActual.contains(actual)) {
+			failIterablesNotEqual(expected, actual, indexes, messageOrSupplier);
+		}
+		else {
+			visitedIterablesExpected.add(expected);
+			visitedIterablesActual.add(actual);
+		}
```

Comment:
Can't this be simplified as follows?

```suggestion
		if (!visitedIterablesExpected.add(expected) && !visitedIterablesActual.add(actual)) {
			failIterablesNotEqual(expected, actual, indexes, messageOrSupplier);
		}
```

Reply:
Good point, will change it (the check must be "or": `!visitedIterablesExpected.add(expected) || !visitedIterablesActual.add(actual)`

Reply:
😁

Reply:
Hi @sbrannen, could you please mark the request as solved. I force pushed the changes (maybe not best idea from my side). Thanks!